sound/soc/sdca/sdca_interrupts.c

```diff
@@ -0,0 +1,467 @@
+	int ret = 0;
+
+	if (!slave->sdca_data.interrupt_info)
+		return -ENODEV;
+
+	interrupt_info = slave->sdca_data.interrupt_info;
+
+	/* check first if the interrupt masks are consistent */
+	hw_register_mask = interrupt_info->supported_hw_register_mask;
+	if (!hw_register_mask) {
+		dev_err(dev,
+			"%s: supported_hw_register mask not initialized\n",
+			__func__);
+		return -EINVAL;
+	}
+
+	if (!source_mask) {
+		dev_err(dev,
+			"%s: source mask not set\n",
+			__func__);
+		return -EINVAL;
+	}
+
+	max_register = __fls(hw_register_mask); /* __fls is zero-based, values should be 0..3 */
+	max_source = __fls(source_mask);     /* __fls is zero-based, values should be 0..31 */
+
+	if (max_source >= (8 << max_register)) {
+		dev_err(dev,
+			"%s: source mask %#x incompatible with supported registers %#lx\n",
+			__func__, source_mask, hw_register_mask);
+		return -EINVAL;
+	}
+
+	/* now start the interrupt mask updates */
+	mutex_lock(&interrupt_info->irqs_lock);
+
+	for_each_set_bit(reg_index, &hw_register_mask, 32) {
+		int shift = reg_index << 3;
+		int source_mask_byte =  (source_mask >> shift) & GENMASK(7, 0);
+
+		if (!source_mask_byte)
+			continue;
+
+		ret = sdca_interrupt_register_mask_rmw(slave, reg_index,
+						       source_mask_byte,
+						       enable);
+		if (ret < 0)
+			goto unlock;
+	}
+
+	/*
+	 * Almost done, keep track of the combined interrupt mask used to
+	 * filter interrupts in the handler.
+	 */
+	if (enable)
+		interrupt_info->enabled_interrupt_mask |= source_mask;
+	else
+		interrupt_info->enabled_interrupt_mask &= ~source_mask;
+
+unlock:
+	mutex_unlock(&interrupt_info->irqs_lock);
+
+	return ret;
+}
+EXPORT_SYMBOL_NS(sdca_interrupt_enable, SND_SOC_SDCA_IRQ_HANDLER);
+
+void sdca_interrupt_clear_history(struct sdw_slave *slave, u32 preserve_mask)
+{
+	struct sdca_interrupt_info *interrupt_info;
+
+	interrupt_info = slave->sdca_data.interrupt_info;
+
+	/*
+	 * Clear all history except for the interrupts set in preserve_mask.
+	 * This is very useful for SDCA UMP processing, where the
+	 * interrupt is only thrown once when the ownership changes to
+	 * HOST. If the processing happens in a work queue, and a new interrupt
+	 * cancels the work queue, the interrupt will not be signaled again
+	 */
+	interrupt_info->detected_interrupt_mask &= preserve_mask;
+}
+EXPORT_SYMBOL_NS(sdca_interrupt_clear_history, SND_SOC_SDCA_IRQ_HANDLER);
+
+/* helper called with the 'irq_lock' mutex held */
+static int sdca_interrupt_register_handler(struct sdw_slave *slave,
+					   struct sdca_interrupt_info *interrupt_info,
+					   int reg_index)
+{
+	struct device *dev = &slave->dev;
+	unsigned long status2;
+	unsigned long status;
+	const int retry = 3;
+	int count = 0;
+	int bit;
+	int ret;
+
+	ret = sdw_read_no_pm(slave, SDW_SCP_SDCA_INT1 + reg_index);
+	if (ret < 0) {
+		dev_err_ratelimited(dev,
+				    "%s: read of SDW_SCP_SDCA_INT%d failed: %d\n",
+				    __func__, reg_index + 1, ret);
+		goto io_error;
+	}
+	status = ret;
```

Comment:
@plbossart When I tested rt712_vb, I will get some bits that is not registered in the registered_source_mask and can't be clear by writing status back. As a result, the while loop will not stop until count == retry. I think we should only check the 
registered source.
Like `status &= interrupt_info->registered_source_mask >> (reg_index << 3);`
This is also needed in the loop.

Reply:
I only added the same BIT_0 and BIT_8 interrupts for the jack, but with VB there's a new function so likely additional interrupts maybe?
we can solve this by doing the filter suggested above, but also make sure we have the proper registered_source_mask.

Reply:
I believe that @shumingfan will set registered_source_mask properly. :)

Reply:
I added a warning if we get an interrupt bit that wasn't registered.